internal/controller/metalmachine_controller.go

```diff
@@ -35,11 +42,101 @@ type MetalMachineReconciler struct {
 // +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
 
 func (r *MetalMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
+
+	// Fetch the MetalMachine.
+	metalMachine := &infrav1.MetalMachine{}
+	err := r.Get(ctx, req.NamespacedName, metalMachine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the Machine.
+	machine, err := util.GetOwnerMachine(ctx, r.Client, metalMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		logger.Info("Machine Controller has not yet set OwnerRef")
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("machine", klog.KObj(machine))
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		logger.Info("Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, metalMachine) {
+		logger.Info("MetalMachine or linked Cluster is marked as paused, not reconciling")
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("cluster", klog.KObj(cluster))
+
+	metalClusterName := client.ObjectKey{
+		Namespace: metalMachine.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+
+	metalCluster := &infrav1.MetalCluster{}
+	if err := r.Client.Get(ctx, metalClusterName, metalCluster); err != nil {
+		logger.Info("MetalCluster is not available yet")
+		return reconcile.Result{}, nil
+	}
+
+	// Create the cluster scope.
+	clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+		Client:         r.Client,
+		Logger:         &logger,
+		Cluster:        cluster,
+		MetalCluster:   metalCluster,
+		ControllerName: "metalcluster",
+	})
+
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create cluster scope: %+v", err)
+	}
+
+	// Create the machine scope
+	machineScope, err := scope.NewMachineScope(scope.MachineScopeParams{
+		Client:       r.Client,
+		Cluster:      cluster,
+		Machine:      machine,
+		MetalCluster: metalCluster,
+		MetalMachine: metalMachine,
+	})
+
+	if err != nil {
+		return reconcile.Result{}, errors.Errorf("failed to create machine scope: %+v", err)
+	}
+
+	// Always close the scope when exiting this function, so we can persist any MetalMachine changes.
+	defer func() {
+		if err := machineScope.Close(); err != nil && err == nil {
+			logger.Error(err, "failed to close MetalMachine scope")
+		}
+	}()
 
-	// TODO(user): your logic here
+	// Return early if the object or Cluster is paused.
+	if annotations.IsPaused(cluster, metalMachine) {
```

Comment:
Let's discuss offline if it's possible to have a self-sufficient `MetalCluster`, `MetalMachine` API that can be deployed/tested independently of the `Machine`, `Cluster`, mainly to achieve the separation of concerns here.